### PR TITLE
Use nonce package to generate verification keys and CSRF tokens

### PR DESCRIPTION
**Pull request by geraldus**
Also removed almost all warnings and applied **stylish-haskell**.

### CODE REVIEW

**Comment by meteficha** [comment]
`nonce` wasn't on Stackage, which would be a problem, but I just sent a PR to fix that. https://github.com/fpco/stackage/pull/632

**Comment by geraldus** [comment]
One way I can create a single generator here is to use `unsafePerormIO`, but I'm not certain is it good practice too.

Also GHC 7.10.1 build failed already, and error message looks odd for me.

> I just sent a PR
Great!

**Comment by meteficha** [comment]
I'm not sure what you mean by using `unsafePerformIO`.  Do you mean creating a global generator via the non-inline `IORef` trick?

**Comment by geraldus** [comment]
@meteficha yeah

```haskell
nonceGenerator = unsafePerformIO Nonce.new
```

**Comment by geraldus** [comment]
@snoyberg sorry for noise, so should I go the way Felipe suggested?

**Comment by snoyberg** [comment]
I'm OK with the `unsafePerformIO` approach in this case, but I don't know the internals of `nonce`. I leave the decision about which way is better to Felipe, who actually knows what he's talking about here (as opposed to me, who is essentially speaking from pure ignorance).

**Comment by meteficha** [comment]
I don't like the `unsafePerformIO` trick in general, but there's no harm in using it here wrt `nonce`.  I think @snoyberg should have the final say regarding `Yesod.Auth.Email`, so I say go ahead :).

**Comment by geraldus** [comment]
@snoyberg @meteficha updated

phew, finally I've found time to make updates

**Comment by geraldus** [comment]
oh, I have to rebase to master first, sorry will do in a moment!

File: yesod-auth/Yesod/Auth/Email.hs [hunk 1]
```diff
@@ -586,3 +586,7 @@ setLoginLinkKey :: (YesodAuthEmail site, MonadHandler m, HandlerSite m ~ site) =
 setLoginLinkKey aid = do
     now <- liftIO getCurrentTime
     setSession loginLinkKey $ TS.pack $ show (toPathPiece aid, now)
+
+
+defaultNonceGen :: Nonce.Generator
+defaultNonceGen = unsafePerformIO (Nonce.new)
```
Comment (meteficha) [inline yesod-auth/Yesod/Auth/Email.hs]: I think you need a `NOINLINE` pragma.

File: yesod-auth/Yesod/Auth/GoogleEmail2.hs [hunk 2]
```diff
@@ -549,3 +550,7 @@ allPersonInfo (A.Object o) = map enc $ M.toList o
     where enc (key, A.String s) = (key, s)
           enc (key, v) = (key, TL.toStrict $ TL.toLazyText $ A.encodeToTextBuilder v)
 allPersonInfo _ = []
+
+
+defaultNonceGen :: Nonce.Generator
+defaultNonceGen = unsafePerformIO (Nonce.new)
```
Comment (meteficha) [inline yesod-auth/Yesod/Auth/GoogleEmail2.hs]: Same here.

**Comment by meteficha** [comment]
After the `NOINLINE` pragmas are added, the `nonce` part of things will look good to me. :+1: 

BTW, @geraldus, did you test it?

**Comment by geraldus** [comment]
Your mean exactly these changes?  No, I didn't.  I can test this a bit later.

**Comment by geraldus** [comment]
@meteficha updated

**Comment by geraldus** [comment]
OK, I've created simple project in sandbox:
```bash
$ mkdir nonce-test
$ cd nonce-test
$ cabal sandbox init
$ cabal sandbox add-source ~/forks/yesod/yesod-core
$ cabal sandbox add-source ~/forks/yesod/yesod-form
$ cabal sandbox add-source ~/forks/yesod/yesod-auth
$ cabal sandbox add-source ~/forks/yesod/yesod-persistent
$ cabal init
```
Here is [example code][ex].

I was able to login using Google API and create new user and verify registration via email.  I don't know better way to test this.

I've faced only one issue related to `nonce` package with one of my projects so far: I use `markdown` package to render verification letters sent to users, there is something like this
```md
Verify your email by clicking on [this link](#{ver_url})
Direct link: #{ver_url}
```
and if there is two underscores in verification key (e.g. YzdCYQN0wsYD_0_GhSf6ZIsY), markdown renders direct link incorrect, italicizing part of it.
[ex]:https://gist.github.com/geraldus/5db4bf42c73ea8035a6b

**Comment by meteficha** [comment]
It seems to me that the presence of underscores is not a big deal.  @geraldus, can you confirm that using the following works for you?

```markdown
Verify your email by clicking on [this link](#{ver_url})
Direct link: `#{ver_url}`
```

@snoyberg, looks good to me.

**Comment by geraldus** [comment]
@meteficha yes, I confirm.

**Comment by snoyberg** [comment]
Awesome, thank you!

**Pull request merged**